File: mobile/lib/domain/usecases/submit_lesson_answer_use_case.dart
import '../../data/services/i_ai_tutor_service.dart';
import '../entities/question.dart';

class LessonAnswerResult {
  final bool isCorrect;
  final String selectedOption;
  final String correctOption;
  final AIExplanationResponse? aiExplanation;

  const LessonAnswerResult({
    required this.isCorrect,
    required this.selectedOption,
    required this.correctOption,
    this.aiExplanation,
  });
}

class SubmitLessonAnswerUseCase {
  final IAITutorService aiTutorService;

  const SubmitLessonAnswerUseCase(this.aiTutorService);

  Future<LessonAnswerResult> execute({
    required Question question,
    required int selectedIndex,
    String? learnerName,
  }) async {
    final isCorrect = selectedIndex == question.correctIndex;
    final selectedOption = question.options[selectedIndex];
    final correctOption = question.options[question.correctIndex];

    if (isCorrect) {
      return LessonAnswerResult(
        isCorrect: true,
        selectedOption: selectedOption,
        correctOption: correctOption,
      );
    }

    // Call AI Tutor service for diagnostic explanation
    final aiExplanation = await aiTutorService.explainMistake(
      AIExplanationRequest(
        courseId: question.courseId,
        skillId: question.skillId,
        questionId: question.id,
        questionText: question.text,
        selectedOption: selectedOption,
        correctOption: correctOption,
        learnerName: learnerName,
      ),
    );

    return LessonAnswerResult(
      isCorrect: false,
      selectedOption: selectedOption,
      correctOption: correctOption,
      aiExplanation: aiExplanation,
    );
  }
}

File: src/domain/usecases/SubmitLessonAnswerUseCase.ts
import { IAITutorService, AIExplanationResponse } from '@/data/services/IAITutorService';
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { Question } from '../entities/Question';

export interface LessonAnswerResult {
  isCorrect: boolean;
  selectedOption: string;
  correctOption: string;
  aiExplanation?: AIExplanationResponse;
}

export class SubmitLessonAnswerUseCase {
  constructor(
    private aiTutorService: IAITutorService,
    private learnerRepo?: ILearnerRepository
  ) {}

  async execute(
    question: Question,
    selectedIndex: number,
    learnerName?: string,
    lessonId?: string
  ): Promise<LessonAnswerResult> {
    const isCorrect = selectedIndex === question.correctIndex;
    const selectedOption = question.options[selectedIndex] || '';
    const correctOption = question.options[question.correctIndex] || '';

    // Log answer attempt if repository provided
    if (this.learnerRepo) {
      await this.learnerRepo.recordAnswerAttempt({
        courseId: question.courseId,
        skillId: question.skillId,
        lessonId: lessonId || 'lesson_general',
        questionId: question.id,
        selectedIndex,
        selectedAnswer: selectedOption,
        isCorrect,
      });
    }

    if (isCorrect) {
      return {
        isCorrect: true,
        selectedOption,
        correctOption,
      };
    }

    // Call AI Tutor service for diagnostic feedback on error
    const aiExplanation = await this.aiTutorService.explainMistake({
      courseId: question.courseId,
      skillId: question.skillId,
      questionId: question.id,
      questionText: question.text,
      selectedOption,
      correctOption,
      learnerName,
    });

    return {
      isCorrect: false,
      selectedOption,
      correctOption,
      aiExplanation,
    };
  }
}
